pennylane/grouping/transformations.py

```diff
@@ -109,15 +109,27 @@ def diagonalize_qwc_grouping(qwc_grouping):
         if isinstance(term, Tensor):
             for sigma in term.obs:
                 if sigma.name != "Identity":
-                    pauli_operators.append(sigma)
+                    if not any(
+                        [
+                            are_identical_pauli_words(sigma, existing_pauli)
+                            for existing_pauli in pauli_operators
+                        ]
+                    ):
+                        pauli_operators.append(sigma)
                     if diag_term is None:
                         diag_term = qml.PauliZ(wires=sigma.wires)
                     else:
                         diag_term @= qml.PauliZ(wires=sigma.wires)
         elif isinstance(term, paulis_with_identity):
             sigma = term
             if sigma.name != "Identity":
-                pauli_operators.append(sigma)
+                if not any(
+                    [
+                        are_identical_pauli_words(sigma, existing_pauli)
+                        for existing_pauli in pauli_operators
+                    ]
+                ):
+                    pauli_operators.append(sigma)
                 diag_term = qml.PauliZ(wires=sigma.wires)
```

Comment:
One way for making Codefactor happy would be to create two small helper functions:
* `get_tensor_diag_term`
* `get_diag_term`
(or comparable names.)

The idea would be to move the checks & building of `diag_term` to these helpers and end up with code similar to
```python
    for term in qwc_grouping:
        diag_term = None
        if isinstance(term, Tensor):
            diag_term = get_tensor_diag_term(term)
        elif isinstance(term, paulis_with_identity):
            diag_term = get_diag_term(term)
        diag_terms.append(diag_term)
```
This overall could make the logic nice to read too.

Reply:
I've added a `diagonalize_pauli_word` function which cleans up `diagonalize_qwc_grouping` a bit. Although, I think the origin of complication in the codeblock you've highlighted are moreso the checks on what's in `pauli_operators`, rather than logic involved in getting the diagonalized Pauli word. Since, currently, redundant Pauli operators can be added to the same set, e.g. `PauliX('a') in set([PauliX('a') , PauliY('b')])` turns out `False`, so things like `set([PauliX('a'), PauliX('a'])` exist w/ 2 elements. `pauli_operators` needs to be a list of non-redundant Pauli operations, otherwise `qwc_rotation` will append additional (over-)rotations. This is the origin of the `if not any` checks in `diagonalize_qwc_grouping` you've highlighted, which seem to reduce readability. These might be alleviated by an additional `is_pauli_word_in_list` auxiliary function. Let me know if this would be helpful. I think an `is_pauli_word_in_list` function, `are_identical_pauli_words`, etc, would all be deprecated/trivialized if an `__eq__` method was made for the `Observable class`, such that e.g. `PauliX('a') == PauliX('a')` is `True`. This might be a more general approach to cleaning up the logic of these different seemingly simple Pauli word checks.